clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -393,17 +340,20 @@ private void throwIfRecordTooLarge() throws RecordTooLargeException {
             return Collections.emptyMap();
         } else {
             Map<TopicPartition, List<ConsumerRecord<K, V>>> drained = new HashMap<>();
-            throwIfOffsetOutOfRange();
-            throwIfUnauthorizedTopics();
-            throwIfRecordTooLarge();
-
             int maxRecords = maxPollRecords;
-            Iterator<PartitionRecords<K, V>> iterator = records.iterator();
-            while (iterator.hasNext() && maxRecords > 0) {
-                PartitionRecords<K, V> part = iterator.next();
-                maxRecords -= append(drained, part, maxRecords);
-                if (part.isConsumed())
-                    iterator.remove();
+            Iterator<FetchCompletion> completedFetchesIterator = completedFetches.iterator();
+
+            while (maxRecords > 0) {
+                if (nextInLineRecords == null || nextInLineRecords.isConsumed()) {
+                    if (!completedFetchesIterator.hasNext())
+                        break;
+
+                    FetchCompletion completion = completedFetchesIterator.next();
+                    completedFetchesIterator.remove();
+                    nextInLineRecords = parseFetchedData(completion);
+                } else {
+                    maxRecords -= append(drained, nextInLineRecords, maxRecords);
```

Comment:
I was looking at the `append` code and it seems a bit brittle. It assumes that:

* The collection returned from `PartitionRecords.take` is safe to mutate even though the latter returns `Collections.emptyList()` if `records == null`
* That `part.take` will always return at least one element

This is fine today, but it may be worth making it a bit more robust to refactorings.